promql/prettier/prettier.go

```diff
@@ -0,0 +1,480 @@
+package prettier
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/rulefmt"
+	"github.com/prometheus/prometheus/promql/parser"
+	"io/ioutil"
+	"reflect"
+	"regexp"
+)
+
+const (
+	// PrettifyRules for prettifying rules files along with the
+	// expressions in them.
+	PrettifyRules = iota
+	// PrettifyExpression for prettifying instantaneous expressions.
+	PrettifyExpression
+)
+
+// Prettier handles the prettifying and formatting operation over a
+// list of rules files or a single expression.
+type Prettier struct {
+	files      []string
+	expression string
+	Type       uint
+	Node       parser.Expr
+	pd         *padder
+}
+
+type padder struct {
+	indent      string
+	columnLimit int
+	// isNewLineApplied represents a new line was added previously, after
+	// appending the item to the result. It should be set to default
+	// after applying baseIndent.
+	isNewLineApplied bool
+	// containsGrouping represents whether a binary expression contains
+	// a grouping modifier.
+	containsGrouping bool
```

Comment:
IIRC we want to line spilt a binary expression independently of whether there is a grouping modifier.

Rather, in all cases where both sides are Vectors, the expression is split.

Reply:
I think, as per the recent comment by Brian, binary expressions should break in these cases:

1. Contains a grouping modifier (even if one child is a scalar).
2. Violates column limit, but no child should be a scalar.

Apart from these, all other cases will require Binary expressions in the same line.
However, I need to confirm these again from design docs to be sure.

Reply:
That's what I thought at first, too.

But later it turned out he actually meant what I wrote above.

Reply:
This is addressed. Hence, resolving.